pkg/daemon/cleanup/cleanup.go

```diff
@@ -0,0 +1,109 @@
+//go:build windows
+
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cleanup
+
+import (
+	"context"
+	"net"
+
+	"github.com/pkg/errors"
+	"golang.org/x/sys/windows/svc"
+	core "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/openshift/windows-machine-config-operator/pkg/daemon/controller"
+	"github.com/openshift/windows-machine-config-operator/pkg/daemon/manager"
+	"github.com/openshift/windows-machine-config-operator/pkg/daemon/winsvc"
+	"github.com/openshift/windows-machine-config-operator/pkg/nodeconfig"
+	"github.com/openshift/windows-machine-config-operator/pkg/servicescm"
+)
+
+// Execute removes an instance from the cluster. If the parameter flag is true, the node object is not deleted.
+// If we are able to get the services ConfigMap tied to the desired version, all services defined in it are cleaned up.
+// TODO: Best effort cleanup if node object, desired version annotation, or services ConfigMap are not attainable
+func Execute(c client.Client, addrs []net.Addr, svcMgr manager.Manager, preserveNode bool) error {
+	ctx := context.TODO()
+	node, err := controller.GetAssociatedNode(c, addrs)
+	if err != nil {
+		klog.Exitf("could not node object associated with this instance: %s", err.Error())
+	}
+	klog.Infof("cleaning up managed services on node %s", node.Name)
+
+	desiredVersion, present := node.Annotations[nodeconfig.DesiredVersionAnnotation]
+	if !present {
+		return errors.Errorf("node missing desired version annotation")
+	}
+	// Fetch the CM of the desired version
+	cm := &core.ConfigMap{}
+	err = c.Get(ctx,
+		client.ObjectKey{Namespace: controller.WmcoNamespace, Name: servicescm.NamePrefix + desiredVersion}, cm)
+	if err != nil {
+		return err
+	}
+	return cleanCMServices(c, ctx, svcMgr, cm, node, preserveNode)
+}
+
+// cleanCMServices removes all the services defined in the given ConfigMap from this instance.
+// The given Node object is deleted if the given flag is false
+func cleanCMServices(c client.Client, ctx context.Context, svcMgr manager.Manager, cm *core.ConfigMap, node *core.Node,
+	preserveNode bool) error {
+	cmData, err := servicescm.Parse(cm.Data)
+	if err != nil {
+		return err
+	}
+	existingSvcs, err := svcMgr.GetExistingServices()
+	if err != nil {
+		return err
+	}
+
+	klog.Infof("removing services defined in ConfigMap %s", cm.Name)
+	// Build up any failures
+	failedRemovals := []string{}
```

Comment:
How would we return/display them in that case? I'm struggling with balancing verbosity and accuracy of the error messages, but think the current approach is the best of both.

Reply:
I am seeing many examples of dealing with errors. There is [hashicorp/go-multierror](https://pkg.go.dev/github.com/hashicorp/go-multierror). Not saying we should use this but using a list of errors and looping through them at the call site might be easier to read rather than a huge concatenated list.

Reply:
Even if we use the multierror library, when WICD errors are printed in the WMCO log, the list of errors will be printed as one giant list. 

Please let me know if the latest version is satisfactory - collect the service names on failed deletion and print them together at the end.

Reply:
Please reply with an example WMCO log output with the error as you have implemented.